File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/TestViewControllers/ScrubberViewController/ThumbnailItemView.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
Custom NSScrubberItemView for displaying an image.
*/

#import "ThumbnailItemView.h"

@interface ThumbnailItemView ()

@property (strong) NSCache<NSString *, NSImage *> *thumbnailCache;
@property (strong) NSImageView *imageView;
@property (strong) NSProgressIndicator *spinner;

@end


#pragma mark -

@implementation ThumbnailItemView

- (instancetype)initWithFrame:(NSRect)frameRect
{
    self = [super initWithFrame:frameRect];
    if (self != nil)
    {
        _thumbnail = [[NSImage alloc] initWithSize:frameRect.size];
        _thumbnailCache = [[NSCache alloc] init];
        _imageView = [NSImageView imageViewWithImage:_thumbnail];
        [self.imageView setAutoresizingMask:(NSAutoresizingMaskOptions)(NSViewWidthSizable | NSViewHeightSizable)];
        
        _spinner = [[NSProgressIndicator alloc] init];
        self.spinner.indeterminate = YES;
        self.spinner.style = NSProgressIndicatorStyleSpinning;
        [self.spinner sizeToFit];
        self.spinner.frame = CGRectInset(self.bounds, (self.bounds.size.width - self.spinner.frame.size.width)/2.0, (self.bounds.size.height - self.spinner.frame.size.height)/2.0);
        self.spinner.hidden = YES;
        self.spinner.controlSize = NSControlSizeSmall;
        self.spinner.appearance = [NSAppearance appearanceNamed:NSAppearanceNameVibrantDark];
        self.spinner.autoresizingMask = (NSAutoresizingMaskOptions)(NSViewMinXMargin | NSViewMaxXMargin | NSViewMinYMargin | NSViewMaxYMargin);
        
        [self addSubview:self.imageView];
        [self addSubview:self.spinner];
    }
    
    return self;
}

- (void)updateLayer
{
    [self.layer setBackgroundColor:[NSColor controlColor].CGColor];
}

- (void)layout
{
    [super layout];
    
    _imageView.frame = self.bounds;
    
    [_spinner sizeToFit];
    _spinner.frame = CGRectInset(self.bounds, (self.bounds.size.width - _spinner.frame.size.width)/2.0, (self.bounds.size.height - _spinner.frame.size.height)/2.0);
}

- (NSImage *)thumbnail
{
    return _imageView.image;
}

- (void)setThumbnail:(NSImage *)thumbnail
{
    _spinner.hidden = YES;
    [_spinner stopAnimation:nil];
    _imageView.hidden = NO;
    _imageView.image = thumbnail;
}

- (NSString *)imageName
{
    return _imageName;
}

- (void)setImageName:(NSString *)imageName
{
    if (_imageName != imageName)
    {
        NSImage *cachedThumbnail = [_thumbnailCache objectForKey:imageName];
        if (cachedThumbnail)
        {
            [self setThumbnail: cachedThumbnail];
            return;
        }
        _spinner.hidden = NO;
        [_spinner startAnimation:nil];
        _imageView.hidden = YES;
        
        NSString *currentName = imageName;
        
        dispatch_async(dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0), ^{

            NSImage *fullImage = [NSImage imageNamed:currentName];
            if (fullImage == nil)
            {
                dispatch_async(dispatch_get_main_queue(), ^{
                    if (currentName == self.imageName) {
                        [self setThumbnail:[[NSImage alloc] initWithSize:NSZeroSize]];
                    }
                });
                return;
            }
            
            NSSize imageSize = [fullImage size];
            CGFloat thumbnailHeight = 30;
            NSSize thumbnailSize = NSMakeSize(ceil(thumbnailHeight * imageSize.width / imageSize.height), thumbnailHeight);
            
            NSImage *thumbnail = [[NSImage alloc] initWithSize:thumbnailSize];
            
            [thumbnail lockFocus];
            [fullImage drawInRect:NSMakeRect(0, 0, thumbnailSize.width, thumbnailSize.height) fromRect:NSMakeRect(0, 0, imageSize.width, imageSize.height) operation:NSCompositingOperationSourceOver fraction:1.0];
            [thumbnail unlockFocus];
            
            [self.thumbnailCache setObject:thumbnail forKey:currentName];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self setThumbnail:thumbnail];
            });
        });
    }
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/ThumbnailItemView.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
Custom NSScrubberItemView class for thumbnail images.
*/

import Cocoa

class ThumbnailItemView: NSScrubberItemView {
    
    static let thumbnailCache = NSCache<NSString, NSImage>()
    
    private let imageView: NSImageView
    private let spinner: NSProgressIndicator
    
    private var thumbnail: NSImage {
        didSet {
            spinner.isHidden = true
            spinner.stopAnimation(nil)
            imageView.isHidden = false
            imageView.image = thumbnail
        }
    }
    
    var imageName: String? {
        didSet {
            guard oldValue != imageName else { return }
            guard let imageName = imageName else {
                imageView.image = nil
                return
            }
            
            if let cachedThumbnail = ThumbnailItemView.thumbnailCache.object(forKey: imageName as NSString) {
                thumbnail = cachedThumbnail
                return
            }
            
            spinner.isHidden = false
            spinner.startAnimation(nil)
            imageView.isHidden = true
            
            let currentName = imageName
            DispatchQueue.global(qos: .background).async {
                guard let fullImage = NSImage(named: currentName) else {
                    DispatchQueue.main.async {
                        if currentName == self.imageName {
                            self.thumbnail = NSImage(size: .zero)
                        }
                    }
                    
                    return
                }
                
                let imageSize = fullImage.size
                let thumbnailHeight: CGFloat = 30
                let thumbnailSize = NSSize(width: ceil(thumbnailHeight * imageSize.width / imageSize.height), height: thumbnailHeight)
                
                let thumbnail = NSImage(size: thumbnailSize)
                thumbnail.lockFocus()
                fullImage.draw(in: NSRect(origin: .zero, size: thumbnailSize),
                               from: NSRect(origin: .zero, size: imageSize),
                               operation: .sourceOver,
                               fraction: 1.0)
                thumbnail.unlockFocus()
                
                ThumbnailItemView.thumbnailCache.setObject(thumbnail, forKey: self.imageName! as NSString)
                
                DispatchQueue.main.async {
                    if currentName == self.imageName {
                        self.thumbnail = thumbnail
                    }
                }
            }
        }
    }
    
    required override init(frame frameRect: NSRect) {
        thumbnail = NSImage(size: frameRect.size)
        imageView = NSImageView(image: thumbnail)
        imageView.autoresizingMask = [.width, .height]
        spinner = NSProgressIndicator()
        
        super.init(frame: frameRect)
        
        spinner.isIndeterminate = true
        spinner.style = .spinning
        spinner.sizeToFit()
        spinner.frame = bounds.insetBy(dx: (bounds.width - spinner.frame.width) / 2, dy: (bounds.height - spinner.frame.height) / 2)
        spinner.isHidden = true
        spinner.controlSize = .small
        spinner.appearance = NSAppearance(named: NSAppearance.Name.vibrantDark)
        spinner.autoresizingMask = [.minXMargin, .maxXMargin, .minYMargin, .maxXMargin]
        
        subviews = [imageView, spinner]
    }
    
    required init?(coder: NSCoder) {
        // The system always creates this particular view class programmatically.
        fatalError("init(coder:) has not been implemented")
    }
    
    override func updateLayer() {
        layer?.backgroundColor = NSColor.controlColor.cgColor
    }
    
    override func layout() {
        super.layout()
        
        imageView.frame = bounds
        spinner.sizeToFit()
        spinner.frame = bounds.insetBy(dx: (bounds.width - spinner.frame.width) / 2, dy: (bounds.height - spinner.frame.height) / 2)
    }
    
}

